algebird-core/src/main/scala/com/twitter/algebird/Window.scala

```diff
@@ -128,17 +143,41 @@ case class WindowMonoid[T](
       while (it.hasNext) {
         queue = (queue ++ it.next.items).takeRight(windowSize)
       }
-      val monT: Monoid[T] = p.join
-      Some(Window(monT.sum(queue), queue))
+      Some(Window(monoid.sum(queue), queue))
     }
 
   def fromIterable(ts: Iterable[T]): Window[T] = {
-    if(ts.size == 0) zero
+    if (ts.isEmpty) zero
     else {
-      val monT: Monoid[T] = p.join
-      val right = ts.toList.takeRight(windowSize)
-      val total = monT.sum(right)
-      Window(total, Queue(right: _*))
+      var queue = Queue.empty[T]
+      var size: Int = 0
+      val it = ts.iterator
+      while (it.hasNext) {
+        // avoid materializing the whole list in memory
+        // at one time
+        queue = queue :+ it.next
+        size = size + 1
+        if (size > windowSize) {
+          queue = queue.tail
+          size = size - 1
+        }
+      }
```

Comment:
It's probably possible to geek out on this even further, but this looks good to me for now.